### ISSUE TITLE
SC2094: assumes files too often

### ISSUE THREAD

**Issue by pixelb** [koalaman/shellcheck#282]
echo old > old
         ^-- SC2094: Make sure not to read and write the same file in the same pipeline.
               ^-- SC2094: Make sure not to read and write the same file in the same pipeline.

echo's operands are not files

**Comment by pixelb** [comment]
This even triggers with variables on verion 0.3.4

    printf '%s\n' $users '' >> users
                                 ^-- SC2094: Make sure not to read and write the same file in the same pipeline.

**Comment by l0b0** [comment]
Is this even possible to implement reliably? You could either reduce this check to only look for "obvious" issues like `command < file > file`, enumerate a bunch of common problem patterns like `grep foo file > file`, or exclude common valid patterns like `echo file > file`.

**Comment by kbulygin** [comment]
Another case:

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

date >date
```

#### Here's what shellcheck currently says:
```
Line 3:
date >date
^-- SC2094: Make sure not to read and write the same file in the same pipeline.
      ^-- SC2238: Redirecting to/from command name instead of file. Did you want pipes/xargs (or quote to ignore)?
      ^-- SC2094: Make sure not to read and write the same file in the same pipeline.

```

**Comment by gblues** [comment]
Using redirection prefix syntax also triggers this shellcheck warning:

```
# This is a simple function that spits out its parameters to stderr before exiting
die() {
  2>& echo "$@"
  exit 1
}
```

**Comment by forthrin** [comment]
The following, always run from a directory other than `~` (which the checker obviously cannot know), triggers a warning, though no files will be read and written to simultaneously.

`git=~/git; for f in "$git"/*/; do git -C "$F"... > git`

That said, a good recommendation here would be something like: _"Avoid identically named variables, directories and executables within the same proximity/scope."_ Is there such a check?

**Comment by martinetd** [comment]
Ran into this with `read -r foo < foo` (assigning variable $foo from ./foo)

Adding ./ to file path, e.g. `read -r foo < ./foo` works around the issue.
